[Parse] Introduce `/.../` regex literals

hamishknight opened the pull request:
Start parsing regex literals with `/.../` delimiters when `-enable-bare-slash-regex` is passed.

rdar://83253726

hamishknight commented:
@swift-ci please test

hamishknight commented:
Source compat was run with it enabled in https://github.com/apple/swift/pull/41767

hamishknight commented:
@xedin @beccadax Mind taking a look at the `DiagnosticEngine` part?

hamishknight commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -1040,24 +1040,40 @@ DiagnosticBehavior DiagnosticState::determineBehavior(const Diagnostic &diag) {
+  ActiveDiagnostic.reset();
+}
+
+void DiagnosticEngine::handleDiagnostic(Diagnostic diag) {
```
I usually try to avoid `&&` parameters unless the only right way of using the API is to move the argument, though yeah maybe in this case it's warranted, I'll update it. Thanks!

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

rintaro commented on include/swift/Parse/Parser.h:
```diff
@@ -1745,8 +1750,17 @@ class Parser {
   ParserResult<Expr>
   parseExprPoundCodeCompletion(Optional<StmtKind> ParentKind);
 
+  UnresolvedDeclRefExpr *makeExprOperator(Token opToken);
```
```suggestion
  UnresolvedDeclRefExpr *makeExprOperator(const Token &opToken);
```

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -1991,12 +1996,103 @@ bool Lexer::tryLexRegexLiteral(const char *TokStart) {
+
+  auto priorState = getStateForBeginningOfToken(NextToken, LeadingTrivia);
+
+  // Re-lex from the given state.
+  restoreState(S);
+
+  // While we restored state, that would have re-advanced the lexer. This is
+  // good in that it's filled in all the interesting properties of the token
+  // (trivia, is on new line, etc), but we need to rewind to re-lex the actual
+  // kind.
+  auto *TokStart = getBufferPtrForSourceLoc(NextToken.getLoc());
+  assert(*TokStart == '/');
+  CurPtr = TokStart + 1;
```
I think this is not ideal, but it's OK.

But, I would just introduce a flag like `ShouldTryRegexLiteralOnSlash` in `Lexer`, then
in `lexImpl()`
```cpp
case '/':
  if (ShouldTryRegexLiteralOnSlash && tryLexRegexLiteral(TokStart)) {
    return;
  }
  ... comment thing ...
  return lexOperatorIdentifier();
```
And here:
```cpp
llvm::SaveAndRestore<bool> Tmp(ShouldTryRegexLiteralOnSlash, true);
restoreState(S);
// 'restoreState()' calls 'lexImpl()' so 'NextToken' should be 'regex_literal' now (if it can be).
```

As for `mustBeRegex` thing, I think you could just check `*getBufferPtrForSourceLoc(S.Loc)`

hamishknight commented:
@swift-ci please test

hamishknight committed:
[DiagnosticEngine] Introduce DiagnosticQueue

This allows us to hold tentative diagnostics
independent of other diagnostic transactions.

hamishknight committed:
[Lexer] Delay token diagnostics

Queue up diagnostics when lexing, waiting until
`Lexer::lex` is called before emitting them. This
allows us to re-lex without having to deal with
previously invalid tokens.

hamishknight committed:
Start using '-enable-bare-slash-regex'

Change the flag to imply
'-enable-experimental-string-processing', and
and update some tests to start using it.

hamishknight committed:
[Parse] Emit error on prefix operator containing `/`

When forward slash regex is enabled, start emitting
an error on prefix operators containing the
`/` character.

hamishknight committed:
[Lexer] Remove `r'...'` lexing logic

This spelling is no longer used.

hamishknight committed:
[Parse] Introduce `/.../` regex literals

Start parsing regex literals with `/.../`
delimiters.

rdar://83253726

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

rintaro reviewed (approved):
Looks good. Thank you!

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
@swift-ci please smoke test Linux
